### PR TITLE
fix nan in eval results

### LINKED ISSUE #1971: TypeError: 'NoneType' object is not iterable
[fail.zip](https://github.com/Microsoft/LightGBM/files/2799905/fail.zip)

Code was compiled for GPU.  Uses hash: d169ac8b7049d3e8345e70288f19eaa7dfb8fe38

Issue has been there for many months, but we noticed it happening for non-trivial data as well so now posting issue with example.

### PR DESCRIPTION
**Pull request by guolinke**
fix #1971

### CODE REVIEW

**Comment by StrikerRUS** [comment]
maybe add a regression test?

**Comment by StrikerRUS** [comment]
I've added `assert`s for certainty. Because I remember we had a test in which `cv` run successfully but there was rubbish in result's content (it was cv for lambdarank, if I\m not mistaken).

File: python-package/lightgbm/callback.py [hunk 1]
```diff
@@ -213,7 +213,7 @@ def _callback(env):
             return
         for i in range_(len(env.evaluation_result_list)):
             score = env.evaluation_result_list[i][2]
-            if cmp_op[i](score, best_score[i]):
+            if best_score_list[i] is None or cmp_op[i](score, best_score[i]):
```
Comment (StrikerRUS) [inline python-package/lightgbm/callback.py]: Is this additional condition for the first iter?
https://github.com/Microsoft/LightGBM/blob/fe77eac5e7c95b46085bc78a527809493c9046d6/python-package/lightgbm/callback.py#L201

Comment (guolinke) [reply python-package/lightgbm/callback.py]: this is to forcely push the score of the first iter.

**Pull request merged**